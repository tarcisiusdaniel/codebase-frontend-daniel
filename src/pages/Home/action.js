import { ACTIONS } from '../../constants';

export function getListOfDevelopers() {
  return (dispatch) => {
    dispatch({ type: ACTIONS.LOADING });

    return setTimeout(() => {
      dispatch({
        type: ACTIONS.LIST_OF_DEVELOPERS_FETCHED,
        data: ['Acus', 'Alif', 'Faezal', 'Oka']
      });
    }, 1000);
  };
}