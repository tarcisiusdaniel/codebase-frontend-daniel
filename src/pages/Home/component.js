import React from 'react';
import PropTypes from 'prop-types';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import ListItem from '../../components/ListItem';
import { IMAGES, ROUTES } from '../../configs';
import { toLowerCase } from '../../utils/text';

export default class Component extends React.Component {
  componentWillMount() {
    this.props.actions.getListOfDevelopers();
  }

  _renderList() {
    const { data } = this.props;
    if (data.length === 0) {
      return (
        <p>No data</p>
      );
    } else {
      return (
        <Grid container justify="center" spacing={24}>
          {
            data.map((item, idx) => (
              <Grid item key={idx} md={1} sm={6} xs={12}>
                <ListItem text={item} url={ROUTES.DEVELOPER(toLowerCase(item))} />
              </Grid>
            ))
          }
        </Grid>
      );
    }
  }

  render() {
    const { classes, isLoading } = this.props;
    if (isLoading) {
      return (
        <Typography variant="display2">Loading...</Typography>
      );
    } else {
      return (
        <section className={classes.container}>
          <img className={classes.logo} src={IMAGES.LOGO} />
          <Typography variant="display2">App is Ready</Typography>
          <Typography className={classes.routeExample}>Route example:</Typography>
          {this._renderList()}
        </section>
      );
    }
  }
}

Component.defaultProps = {
  actions: {},
  classes: {},
  data: [],
  isLoading: false,
};

Component.propTypes = {
  actions: PropTypes.object,
  classes: PropTypes.object,
  data: PropTypes.array,
  isLoading: PropTypes.bool,
};