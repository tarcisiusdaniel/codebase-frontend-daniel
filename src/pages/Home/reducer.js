import { ACTIONS } from '../../constants';

const initialState = {
  isLoading: false,
  data: []
};

export default function reducer(state = initialState, action) {
  const { type, data } = action;
  switch (type) {
    case ACTIONS.LOADING:
      return {
        ...state,
        isLoading: true
      };
    case ACTIONS.LIST_OF_DEVELOPERS_FETCHED:
      return {
        ...state,
        isLoading: false,
        data
      };
    default:
      return state;
  }
}