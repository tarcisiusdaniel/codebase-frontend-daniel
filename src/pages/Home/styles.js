const styles = {
  container: {
    textAlign: 'center'
  },
  logo: {
    maxWidth: '100%'
  },
  routeExample: {
    marginBottom: '20px'
  }
};

export default styles;