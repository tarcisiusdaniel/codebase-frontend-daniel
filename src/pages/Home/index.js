import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core/styles';
import Component from './component';
import * as actions from './action';
import styles from './styles';

function mapStateToProps(state) {
  const { isLoading, data } = state.home;
  return {
    isLoading,
    data
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(actions, dispatch)
  };
}

const Styled = withStyles(styles)(Component);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Styled);