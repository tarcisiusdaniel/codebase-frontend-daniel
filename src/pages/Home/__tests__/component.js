import React from 'react';
import renderer from 'react-test-renderer';
import Home from '../component';

jest.mock('../../../components/ListItem');
jest.mock('../../../utils/text');

describe('Home component', () => {
  let props = {};
  beforeEach(() => {
    props = {
      actions: {
        getListOfDevelopers: jest.fn()
      },
      classes: {},
      isLoading: false,
      data: []
    };
  });

  it('renders correctly when loading', () => {
    props.isLoading = true;
    const tree = renderer
      .create(<Home {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when data is ready', () => {
    props.data = ['1', '2', '3'];
    const tree = renderer
      .create(<Home {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  
  it('renders correctly when data is empty', () => {
    const tree = renderer
      .create(<Home {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});