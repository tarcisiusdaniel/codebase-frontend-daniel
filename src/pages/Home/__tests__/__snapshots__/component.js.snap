// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component renders correctly when data is empty 1`] = `
<section>
  <img
    src="/assets/logo.png"
  />
  <h1
    className="MuiTypography-root-1 MuiTypography-display2-4"
  >
    App is Ready
  </h1>
  <p
    className="MuiTypography-root-1 MuiTypography-body1-10"
  >
    Route example:
  </p>
  <p>
    No data
  </p>
</section>
`;

exports[`Home component renders correctly when data is ready 1`] = `
<section>
  <img
    src="/assets/logo.png"
  />
  <h1
    className="MuiTypography-root-1 MuiTypography-display2-4"
  >
    App is Ready
  </h1>
  <p
    className="MuiTypography-root-1 MuiTypography-body1-10"
  >
    Route example:
  </p>
  <div
    className="MuiGrid-container-26 MuiGrid-spacing-xs-24-50 MuiGrid-justify-xs-center-43"
  >
    <div
      className="MuiGrid-item-27 MuiGrid-grid-xs-12-66 MuiGrid-grid-sm-6-74 MuiGrid-grid-md-1-83"
    />
    <div
      className="MuiGrid-item-27 MuiGrid-grid-xs-12-66 MuiGrid-grid-sm-6-74 MuiGrid-grid-md-1-83"
    />
    <div
      className="MuiGrid-item-27 MuiGrid-grid-xs-12-66 MuiGrid-grid-sm-6-74 MuiGrid-grid-md-1-83"
    />
  </div>
</section>
`;

exports[`Home component renders correctly when loading 1`] = `
<h1
  className="MuiTypography-root-1 MuiTypography-display2-4"
>
  Loading...
</h1>
`;
